File: macos/WayforkCore/Sources/WayforkCore/Control/LogQuery.swift
import Foundation

/// `wayforkctl logs` filters (docs/design/09-wayforkctl.md § logs): source, level threshold,
/// case-insensitive substrings, a start time and a tail length. Pure; the CLI only finds the
/// files.
public struct LogQuery: Sendable, Hashable {
    /// Empty: every source. `openvpn` stands for every `openvpn:<id>`.
    public var sources: [String]
    /// Lines at this level or more severe.
    public var level: LogLevel
    /// Every one must occur in the message, case-insensitively.
    public var grep: [String]
    public var since: Date?
    /// The last `tail` matching lines; 0 means no cap.
    public var tail: Int

    public static let defaultTail = 100

    public init(
        sources: [String] = [],
        level: LogLevel = .debug,
        grep: [String] = [],
        since: Date? = nil,
        tail: Int = LogQuery.defaultTail
    ) {
        self.sources = sources
        self.level = level
        self.grep = grep
        self.since = since
        self.tail = tail
    }

    public func matches(_ line: LogLine) -> Bool {
        guard line.level <= level else { return false }
        if let since, line.ts < since { return false }
        if !sources.isEmpty, !sources.contains(where: { Self.source(line.source, matches: $0) }) {
            return false
        }
        return grep.allSatisfy { line.message.range(of: $0, options: .caseInsensitive) != nil }
    }

    /// Filters every stream, merges them by timestamp (stable for equal stamps) and keeps
    /// the tail.
    public func run(_ streams: [[LogLine]]) -> [LogLine] {
        var merged: [(index: Int, line: LogLine)] = []
        for stream in streams {
            for line in stream where matches(line) {
                merged.append((merged.count, line))
            }
        }
        merged.sort { ($0.line.ts, $0.index) < ($1.line.ts, $1.index) }
        let lines = merged.map(\.line)
        guard tail > 0, lines.count > tail else { return lines }
        return Array(lines.suffix(tail))
    }

    static func source(_ source: String, matches filter: String) -> Bool {
        let filter = filter.lowercased()
        let source = source.lowercased()
        return source == filter || (filter == "openvpn" && source.hasPrefix("openvpn:"))
    }

    /// `90s`, `15m`, `2h`, `1d` before `now`, or an ISO-8601 timestamp (with or without
    /// fractional seconds). Nil when the text is neither.
    public static func parseSince(_ text: String, now: Date = Date()) -> Date? {
        let text = text.trimmingCharacters(in: .whitespaces)
        if let unit = text.last, let amount = Double(text.dropLast()), amount >= 0 {
            let seconds: Double? =
                switch unit {
                case "s": 1
                case "m": 60
                case "h": 3600
                case "d": 86_400
                default: nil
                }
            if let seconds { return now.addingTimeInterval(-amount * seconds) }
        }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: text) { return date }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: text)
    }
}

/// Reads `runtime.log` / `wayfork.log` and, when needed, their rotated siblings, newest file
/// first, stopping as soon as older files cannot contribute.
public enum LogArchive {
    /// Lines of `<name>.log` and `<name>-<stamp>.log` in `directory`, oldest first, that
    /// match `query` — enough of them for `query.tail` (every one when `tail` is 0).
    public static func lines(directory: URL, name: String, query: LogQuery) -> [LogLine] {
        var files = AppLogFile.rotatedFiles(directory: directory, name: name)
        files.append(directory.appendingPathComponent("\(name).log"))
        var chunks: [[LogLine]] = []
        var count = 0
        for file in files.reversed() {
            guard let text = try? String(contentsOf: file, encoding: .utf8) else { continue }
            let parsed = text.split(separator: "\n").compactMap { LogLineFormat.parse(String($0)) }
            let matching = parsed.filter(query.matches)
            chunks.append(matching)
            count += matching.count
            if let since = query.since, let first = parsed.first, first.ts < since { break }
            if query.since == nil, query.tail > 0, count >= query.tail { break }
        }
        return chunks.reversed().flatMap { $0 }
    }
}

/// Replaces configured server addresses in log messages with `server-N` placeholders
/// (numbered in store order). Only exact, whole-token matches: a host inside a longer name
/// or an address inside a longer one stays as it is.
public struct LogRedactor: Sendable {
    private let replacements: [(needle: String, placeholder: String)]

    public init(servers: [String]) {
        var seen: [String: String] = [:]
        var ordered: [(String, String)] = []
        for server in servers {
            let key = server.lowercased()
            guard !key.isEmpty, seen[key] == nil else { continue }
            let placeholder = "server-\(seen.count + 1)"
            seen[key] = placeholder
            ordered.append((key, placeholder))
        }
        // Longest first, so `a.example.com` is replaced before `example.com` could be.
        replacements = ordered.sorted { $0.0.count > $1.0.count }
    }

    /// The addresses of every tunnel's servers, in store order.
    public init(store: Store) {
        self.init(servers: store.tunnels.flatMap(\.kind.serverHosts))
    }

    public var isEmpty: Bool { replacements.isEmpty }

    /// The redacted message and how many addresses were replaced.
    public func redact(_ message: String) -> (String, Int) {
        guard !replacements.isEmpty else { return (message, 0) }
        var result = message
        var count = 0
        for (needle, placeholder) in replacements {
            var searchStart = result.startIndex
            while let range = result.range(
                of: needle, options: .caseInsensitive, range: searchStart..<result.endIndex)
            {
                if Self.isBoundary(result, before: range.lowerBound)
                    && Self.isBoundary(result, after: range.upperBound)
                {
                    let offset = result.distance(from: result.startIndex, to: range.lowerBound)
                    result.replaceSubrange(range, with: placeholder)
                    count += 1
                    searchStart = result.index(
                        result.startIndex, offsetBy: offset + placeholder.count)
                } else {
                    searchStart = range.upperBound
                }
            }
        }
        return (result, count)
    }

    private static func isTokenCharacter(_ character: Character) -> Bool {
        character.isLetter || character.isNumber || character == "." || character == "-"
            || character == "_"
    }

    private static func isBoundary(_ text: String, before index: String.Index) -> Bool {
        guard index > text.startIndex else { return true }
        return !isTokenCharacter(text[text.index(before: index)])
    }

    private static func isBoundary(_ text: String, after index: String.Index) -> Bool {
        guard index < text.endIndex else { return true }
        let next = text[index]
        // A trailing dot ends a sentence, not a name: `server.example.com.` still matches.
        if next == "." {
            let afterDot = text.index(after: index)
            return afterDot == text.endIndex || !isTokenCharacter(text[afterDot])
        }
        return !isTokenCharacter(next)
    }
}

File: windows/service/internal/core/logquery.go
package core

import (
	"sort"
	"strconv"
	"strings"
	"time"
)

// LogQuery is `wayforkctl logs`' filter (F21, docs/design/09-wayforkctl.md § Commands
// (Windows)): source, level threshold, case-insensitive substrings, a start time and a
// tail length — the same semantics as the macOS `LogQuery`.
type LogQuery struct {
	// Sources: empty means every source; `openvpn` stands for every `openvpn:<id>`.
	Sources []string
	// Level: lines at this level or more severe; empty means debug (everything).
	Level LogLevel
	// Grep: every entry must occur in the message, case-insensitively.
	Grep []string
	// Since: the zero time means no limit.
	Since time.Time
	// Tail keeps the last Tail matching lines; 0 means no cap.
	Tail int
}

// DefaultLogTail is `wayforkctl logs`' default --tail.
const DefaultLogTail = 100

func logLevelRank(level LogLevel) int {
	switch level {
	case LogLevelError:
		return 0
	case LogLevelWarning:
		return 1
	case LogLevelInfo:
		return 2
	default:
		return 3
	}
}

// Matches reports whether line passes every filter.
func (q LogQuery) Matches(line LogLine) bool {
	if q.Level != "" && logLevelRank(line.Level) > logLevelRank(q.Level) {
		return false
	}
	if !q.Since.IsZero() && line.TS.Before(q.Since) {
		return false
	}
	if len(q.Sources) > 0 {
		found := false
		for _, source := range q.Sources {
			if logSourceMatches(line.Source, source) {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	message := strings.ToLower(line.Message)
	for _, needle := range q.Grep {
		if !strings.Contains(message, strings.ToLower(needle)) {
			return false
		}
	}
	return true
}

func logSourceMatches(source, filter string) bool {
	source, filter = strings.ToLower(source), strings.ToLower(filter)
	return source == filter || (filter == "openvpn" && strings.HasPrefix(source, "openvpn:"))
}

// Run filters lines, orders them by timestamp (stable for equal stamps) and keeps the
// tail.
func (q LogQuery) Run(lines []LogLine) []LogLine {
	matching := make([]LogLine, 0, len(lines))
	for _, line := range lines {
		if q.Matches(line) {
			matching = append(matching, line)
		}
	}
	sort.SliceStable(matching, func(i, j int) bool { return matching[i].TS.Before(matching[j].TS.Time) })
	if q.Tail > 0 && len(matching) > q.Tail {
		matching = matching[len(matching)-q.Tail:]
	}
	return matching
}

// ParseLogSince accepts `90s`, `15m`, `2h`, `1d` before now, or an RFC 3339 timestamp.
func ParseLogSince(text string, now time.Time) (time.Time, bool) {
	text = strings.TrimSpace(text)
	if len(text) >= 2 {
		units := map[byte]time.Duration{'s': time.Second, 'm': time.Minute, 'h': time.Hour, 'd': 24 * time.Hour}
		if unit, ok := units[text[len(text)-1]]; ok {
			if amount, err := strconv.ParseFloat(text[:len(text)-1], 64); err == nil && amount >= 0 {
				return now.Add(-time.Duration(amount * float64(unit))), true
			}
		}
	}
	if ts, err := time.Parse(time.RFC3339Nano, text); err == nil {
		return ts, true
	}
	return time.Time{}, false
}

// LogSourceForFile maps a log file stem (`sing-box`, `openvpn-<id>`, `daemon`) to the
// source name of its lines.
func LogSourceForFile(stem string) string {
	if rest, ok := strings.CutPrefix(stem, "openvpn-"); ok {
		return "openvpn:" + rest
	}
	return stem
}

// ParseLogFileLine reads one `FormatLogLine` line; false for anything else.
func ParseLogFileLine(source, text string) (LogLine, bool) {
	parts := strings.SplitN(text, " ", 3)
	if len(parts) < 2 {
		return LogLine{}, false
	}
	ts, err := time.Parse(time.RFC3339Nano, parts[0])
	if err != nil {
		return LogLine{}, false
	}
	level, err := ParseLogLevel(strings.ToLower(parts[1]))
	if err != nil || parts[1] != strings.ToUpper(parts[1]) {
		return LogLine{}, false
	}
	message := ""
	if len(parts) == 3 {
		message = parts[2]
	}
	return LogLine{TS: NewTimestamp(ts), Source: source, Level: level, Message: message}, true
}
